include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -544,58 +544,40 @@ __pattern_minmax_element(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _
 // adjacent_find
 //------------------------------------------------------------------------
 
-template <typename _BackendTag, typename _ExecutionPolicy, typename _Iterator, typename _BinaryPredicate>
+template <typename _BackendTag, typename _ExecutionPolicy, typename _Iterator, typename _BinaryPredicate, typename _OrFirstTag>
 _Iterator
 __pattern_adjacent_find(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last,
-                        _BinaryPredicate __predicate, oneapi::dpl::__internal::__or_semantic)
+                        _BinaryPredicate __pred, _OrFirstTag __is_or_semantic)
 {
-    if (__last - __first < 2)
+    const auto __n = __last - __first;
+    if (__n < 2)
         return __last;
 
-    using _Predicate = oneapi::dpl::unseq_backend::single_match_pred<adjacent_find_fn<_BinaryPredicate>>;
+    using _Predicate = oneapi::dpl::unseq_backend::single_match_pred<_BinaryPredicate>;
+
+    auto __keep = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
+    auto __buf = __keep(__first, __last); //a scope lifetime of this instance should be.
+
+    auto __view = __buf.all_view();
+    auto __view1 = oneapi::dpl::__ranges::take_view_simple(__view, __view.size() - 1);
+    auto __view2 = oneapi::dpl::__ranges::drop_view_simple(__view, 1);
 
-    auto __keep1 = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
-    auto __buf1 = __keep1(__first, __last - 1);
-    auto __keep2 = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
-    auto __buf2 = __keep2(__first + 1, __last);
+    using __size_calc = oneapi::dpl::__par_backend_hetero::__min_size_calc;
```

Comment:
Let's use `__first_size_calc` in this particular case: we know that `__view1` and `__view2` have equal sizes, so there no need to call std::min.